Update language around setting Kubernetes pod name in container info (#701)
elastic/apm: specs/agents/metadata.md

    If there is a match to either expression, the capturing group contains the pod ID. We then unescape underscores 
    (`_`) to hyphens (`-`) in the pod UID.

>     Finally, we record the basename as the container ID without any further checks.

Reviewer: Doesn't this change the intent?  Currently, at least in the Node.js APM agent, we only set a value for the Kubernetes pod name (for `metadata.system.kubernetes.pod.name`) in any of the following conditions:
- If the `KUBERNETES_POD_NAME` envvar is set, then use that value.
- If the `KUBERNETES_POD_UID` envvar is set, then use the hostname.
- If a k8s pod UID is parsed from a `/proc/self/cgroup` line with the steps above, then use the hostname.

So far, the changes in the PR make it sound like we would always set `metadata.system.kubernetes.pod.name = hostname`.  Did I misread?

Author: This part of the spec says that we set the pod name to the `KUBERNETES_POD_NAME` env variable, if it's set. Is there something that's not clear about the sentence?

> Sometimes the `KUBERNETES_POD_NAME` is set using the [Downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/),
> so we set the pod name to its value if it exists. Otherwise, we set the pod name to the hostname, because that's the default in Kubernetes.

Reviewer: Assume that no `KUBERNETES_` envvars are set. My read of the changes in this PR is that the spec would suggest that agents *always* set `metadata.system.kubernetes.pod.name` to the hostname.  However, my read of the current spec is that `metadata.system.kubernetes.pod.name` should only be set "If we match a pod UID" in a "/proc/self/cgroup" line.

Author: Thanks for clarifying; I think I've restored the original logic, can you check the latest change?
Also, where in the spec did you get the condition `If the KUBERNETES_POD_UID envvar is set, then use the hostname.`?

 If we match, then the basename is assumed to be a container ID.

Sometimes the `KUBERNETES_POD_NAME` is set using the [Downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/),

> so we set the pod name to its value if it exists. Otherwise, we set the pod name to the hostname, because by default, Kubernetes will set the hostname to the pod name.

Author: ```suggestion
so we set the pod name to its value if it exists. Otherwise, we set the pod name to the hostname, because that's the default value Kubernetes uses.
```

> so we set the pod name to its value if it exists. Otherwise, we set the pod name to the hostname, because by default, Kubernetes will set the hostname to the pod name.

Reviewer 2: ```suggestion
so we set the pod name to its value if it exists. Otherwise, we set the pod name to the hostname, because that's the default in Kubernetes.
```

> In a similar manner, you can inform the agent of the node name and namespace, using the environment variables `KUBERNETES_NODE_NAME` and `KUBERNETES_NAMESPACE`.

Reviewer: nit: Perhaps also mention `KUBERNETES_POD_UID` to set/override `metadata.system.kubernetes.pid.uid`? Though that's adding scope. :)

Author: I added the `KUBERNETES_POD_UID` to the spec. I don't know why it wasn't mentioned, unless I'm missing some special logic?

Reviewer: > I don't know why it wasn't mentioned, unless I'm missing some special logic?

I think it was just overlooked when added. AFAICT all the agents implement reading `KUBERNETES_POD_UID`.

> Sometimes the `KUBERNETES_POD_NAME` is set using the [Downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/),
> so we set the pod name to its value if it exists. Otherwise, we set the pod name to the hostname, because that's the default in Kubernetes.
> In a similar manner, you can inform the agent of the node name and namespace, using the environment variables `KUBERNETES_NODE_NAME` and `KUBERNETES_NAMESPACE`.

Reviewer 3: Do the spec care about the Downward API? Maybe this is simpler?

```suggestion
Users can override the pod name, node name, and namespace using the `KUBERNETES_POD_NAME`, `KUBERNETES_NODE_NAME`, and `KUBERNETES_NAMESPACE` environment variables.
```

Author: I would guess we don't have to mention the Downward API but I don't see much harm in mentioning it, in case a future spec reader wonders how those env variables get set in the first place.
